staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,570 @@
+	}
+	return total
+}
+
+// chooseQueueIndexLocked uses shuffle sharding to select a queue index
+// using the given hashValue and the shuffle sharding parameters of the queueSet.
+func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64) int {
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// DesiredNum is used here instead of numQueues to omit quiescing queues
+	qs.dealer.Deal(hashValue, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// updateQueueVirtualStartTime updates the virtual start time for a queue
+// this is done when a new request is enqueued.  For more info see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching
+func (qs *queueSet) updateQueueVirtualStartTime(request *fq.Request, queue *fq.Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.Requests) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.Requests) == 1 && queue.RequestsExecuting == 0 {
+		// the queue’s virtual start time is set to the virtual time.
+		queue.VirtualStart = qs.virtualTime
+	}
+}
+
+// enqueues a request into an queueSet
+func (qs *queueSet) enqueue(request *fq.Request) {
+	queue := request.Queue
+	queue.Enqueue(request)
+	qs.updateQueueVirtualStartTime(request, queue)
+	qs.numRequestsEnqueued++
+
+	metrics.UpdateFlowControlRequestsInQueue(qs.config.Name, qs.numRequestsEnqueued)
+}
+
+// rejectOrEnqueueLocked rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *queueSet) rejectOrEnqueueLocked(request *fq.Request) bool {
+	queue := request.Queue
+	curQueueLength := len(queue.Requests)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.getRequestsExecutingLocked() >= qs.config.ConcurrencyLimit &&
+		curQueueLength >= qs.config.QueueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(request)
+	return true
+}
+
+func (qs *queueSet) roundRobinQueue() int {
+	qs.robinIndex = (qs.robinIndex + 1) % len(qs.queues)
+	return qs.robinIndex
+}
+
+// selectQueue selects the minimum virtualFinish time from the set of queues
+// the starting queue is selected via roundrobin
+func (qs *queueSet) selectQueue() *fq.Queue {
+	minVirtualFinish := math.Inf(1)
+	var minQueue *fq.Queue
+	var minIndex int
+	for range qs.queues {
+		index := qs.roundRobinQueue()
+		queue := qs.queues[index]
+		if len(queue.Requests) != 0 {
+			currentVirtualFinish := queue.GetVirtualFinish(0, qs.estimatedServiceTime)
+			if currentVirtualFinish < minVirtualFinish {
+				minVirtualFinish = currentVirtualFinish
+				minQueue = queue
+				minIndex = index
+			}
+		}
+	}
+	qs.robinIndex = minIndex
+	return minQueue
+}
+
+// dequeue dequeues a request from the queueSet
+func (qs *queueSet) dequeue() (*fq.Request, bool) {
```

Comment:
+"Locked"?

Reply:
I notice that you did not mention this on all internal methods that (a) must be called with the lock held and (b) do not have "Locked" in their name.  What is your opinion on how to mark locking constraints on methods?

Reply:
Updated

Reply:
The reason I didn't comment everywhere is simple: I reviewed in chunks and especially at first wasn't aware that there was a lock that some things needed to hold. It's best if the convention is universally applied.